src/posts/en/javascript-iterables-and-iterators.md

```diff
@@ -0,0 +1,294 @@
+};
+for (let item of iterable) {
+	console.log(item); // 'a', 'b', 'c'
+}
+```
+
+Now you may be wondering: **"How do I implement custom iteration behavior?"**<br>
+As you already know, adding a `[Symbol.iterator]` method to a given object makes it iterable, but here is the catch: the `[Symbol.iterator]` method must return an *iterator object*, which is actually responsible for the iteration logic. We will cover this in the next section.
+
+Before we move on to iterators, let's recap what the Iterable interface is about:
+
+- It signals that a given object is iterable to consumers (iterable object: "Hey! I have a `[Symbol.iterator]` method.");
+- It provides a standardized way to iterate any iterable object (iterable object: "Call my `[Symbol.iterator]` method if you want an iterator that implements my default iteration logic").
+
+The points above may seem slightly redundant if you are familiar with interfaces (Java, C#), but it is worth being clear as this is a new addition to the JavaScript language.
+
+Note that several built-in functions and language constructs that expect a sequence of values accept iterable objects. For instance, `Promise.all()` accepts any iterable object as argument, not just plain arrays. The same holds true for destructuring assignments (`[a, b] = iterable`) and `yield* iterable`.
+
+## Iterators
+
+Iterator objects are objects that implement the [Iterator interface](http://www.ecma-international.org/ecma-262/6.0/index.html#sec-iterator-interface). That is, iterator objects must have a `next` method that returns a result object in the `{ value: Any, done: Boolean }` format. The first call to `.next()` returns the result of the first iteration (e.g. the 0th item in an array). The `done` property signals when the iterator has been exhausted and no more values are available.
+
+Here is some code to illustrate iterators:
+
+```javascript
+let iterable = ['a', 'b', 'c'];
+
+// Explicit "low-level" iterator consumption:
+let iterator = iterable[Symbol.iterator]();
+iterator.next(); // { value: 'a', done: false }
+iterator.next(); // { value: 'b', done: false }
+iterator.next(); // { value: 'c', done: false }
+iterator.next(); // { value: undefined, done: true }
+
+// for-of abstracts away iterator instantiation and consumption:
+for (let item of iterable) {
+	console.log(item); // 'a', 'b', 'c'
+}
+
+// A crude illustration of for-of's inner workings:
+for (let _iterator = iterable[Symbol.iterator](), _result, item; _result = _iterator.next(), item = _result.value, !_result.done;) {
+	console.log(item); // 'a', 'b', 'c'
+}
+```
+
+Now that we have understanding of iterables and iterators, we can create our own iterable objects and customize their default iteration behavior. Let's implement an iterable object whose iteration behavior provides ten random values:
+
+```javascript
+let iterable = {
+	[Symbol.iterator]() {
+		let index = 0;
+		return {
+			next() {
+				let done = index++ === 10;
```

Comment:
This line relies on side effects. For people learning JS I think it might be better to be more explicit, at the cost of an extra line of code. 

```
index++;
let done = index === 10;
```